market_analysis/views/default.py

```diff
@@ -0,0 +1,450 @@
+
+        except (AttributeError, UnmappedInstanceError):
+            msg = 'Failed: tried to remove a stock that is not in the'\
+                ' portfolio.'
+    else:
+        msg = 'Failed: improper request.'
+    return {'msg': msg}
+
+
+@view_config(route_name='portfolio', renderer="../templates/portfolio.jinja2",
+             permission='secret')
+def portfolio(request):
+    """
+    The main user portfolio page, displays a list of
+    their stocks and a graph. Update number of shares.
+    """
+    current_user_id = request.dbsession.query(Users).filter(
+                Users.username == request.authenticated_userid
+            ).first().id
+    query = request.dbsession.query(Users).\
+        filter(Users.id == current_user_id).first()
+    query = query.children
+
+    if request.method == 'POST':
+        updated_amount = request.POST['amount']
+        for item, val in request.POST.items():
+            if val == 'Update':
+                updated_stock = item
+        current_stock_id = request.dbsession.query(Stocks)\
+            .filter(Stocks.symbol == updated_stock).first().id
+        request.dbsession.query(Association)\
+            .filter(and_(current_stock_id == Association.stock_id,
+                    Association.user_id == current_user_id))\
+            .update({Association.shares: updated_amount})
+
+    list_of_stock_ids = []
+    for row in query:
+        list_of_stock_ids.append(row.child.symbol)
+    if len(list_of_stock_ids) == 0:
+        return HTTPFound(location=request.route_url('search'))
+    else:
+        elements = []
+        for stock in list_of_stock_ids:
+            elements.append({'Symbol': str(stock),
+                             'Type': 'price', 'Params': ['c']})
+        return build_graph(request, elements, True)
+
+
+def package_data(info, entries, msg):
+    """Package data for single_stock_details."""
+    try:
+        info['info'] = entries
+        info['msg'] = msg
+    except TypeError:
+        info = {}
+        info['info'] = entries
+        msg = 'Trouble connecting to API.'
+        info['msg'] = msg
+    return info
+
+
+def check_bad_msg(entries):
+    """Return True if bad request."""
+    if 'Message' in entries.keys():
+        return True
+    return False
+
+
+@view_config(route_name='details', renderer="../templates/details.jinja2",
+                        permission="secret")
+def single_stock_details(request):
+    """Details for single-stock."""
+    entries = {}
+    elements = []
+    msg = ''
+    sym = request.matchdict['sym']
+    resp = requests.get('http://dev.markitondemand.com/'
+                        'Api/v2/Quote/json?symbol=' + sym)
+    if resp.status_code == 200:
+        entries = {key: value for key, value in resp.json().items()}
+        if check_bad_msg(entries):
+            msg = 'Bad request.'
+        elements.append({'Symbol': str(sym), 'Type': 'price', 'Params': ['c']})
+    else:
+        msg = 'Could not fulfill the request.'
```

Comment:
If you had a separate component, perhaps a function or something like that which was responsible for making API requests and returning formatted data, then that could be tested without having to use an entire view to do the testing.  A separate component could also then be moved into a different location, like an out-of-band processor.  

Properly encapsulating this type of action is really helpful in future-proofing your application.